Hammerfest.Server/Dns/HostsFile.cs

```diff
@@ -9,69 +8,82 @@ namespace Hammerfest.Server.Dns;
+    ///     and the value is considered its IP address.
+    /// </returns>
+    /// <remarks>
+    ///     Null as the entries' value indicates that the value is to be deleted.
+    ///     When new entry is added (not replaced) its resulting value in the returned dictionary is null. It is needed
+    ///     to reuse the method on the rollback.
+    /// </remarks>
+    public static Dictionary<string, string?> ModifyEntries(ISystemEnvironment env, IReadOnlyDictionary<string, string?>
+        entriesToSet)
     {
-        foreach (var line in File.ReadAllLines(env.HostsFilePath))
-        {
-            if (ParseLine(line) is (var ipAddress, var currentHost) && currentHost == host)
-                return ipAddress;
-        }
-
-        return null;
-    }
+        if (entriesToSet.Count == 0) return new Dictionary<string, string?>();
 
-    public static void AddEntry(ISystemEnvironment env, string ipAddress, string host)
-    {
-        var content = File.ReadAllText(env.HostsFilePath);
-        var result = new StringBuilder(content);
-        if (content.Length > 0 && !content.EndsWith('\n'))
-            result.Append('\n');
-        result.Append($"{ipAddress} {host}\r\n");
-
-        File.WriteAllText(env.HostsFilePath, result.ToString());
-    }
-
-    public static bool RemoveEntry(ISystemEnvironment env, string host)
-    {
+        var newEntries = new Dictionary<string, string?>(entriesToSet);
         var originalLines = File.ReadAllLines(env.HostsFilePath);
         var resultLines = new List<string>();
-        var removed = false;
+        var originalEntries = new Dictionary<string, string?>();
+
         foreach (var line in originalLines)
         {
-            if (ParseLine(line) is var (_, currentHost) && currentHost == host)
+            var parsedLine = ParseLine(line);
+            switch (parsedLine)
             {
-                removed = true;
+                case var (oldIp, host) when IsAlreadyPresent(host, oldIp):
+                    resultLines.Add($"{newEntries[host]} {host}");
+                    newEntries.Remove(host);
+                    break;
+                case var (oldIp, host) when IsReplacement(host):
+                    resultLines.Add($"{newEntries[host]} {host}");
+                    originalEntries.Add(host, oldIp);
+                    newEntries.Remove(host);
+                    break;
+                case var (oldIp, host) when IsDeletion(host):
+                    originalEntries.Add(host, oldIp);
+                    newEntries.Remove(host);
+                    break;
+                default:
+                    resultLines.Add(line);
+                    break;
             }
-            else
-                resultLines.Add(line);
         }
 
-        if (removed)
+        foreach (var (host, ip) in newEntries) // Only the additions remain here at this point.
+        {
+            resultLines.Add($"{ip} {host}");
+            originalEntries.Add(host, null);
+        }
```

Comment:
Ha, you forget to add a null filter here.

So, if some entry didn't exist in the `hosts` file before, and the user asked us to _delete_ it, then this code will add an empty host entry to the file.

Please fix the bug and add a test case for it.